csrc/runtime/allocations.cpp

```diff
@@ -252,156 +230,77 @@ void fillTensorWithNan(at::Tensor& t) {
   }
 }
 
-at::Tensor allocateTensor(
-    const GlobalBufferInfo& out_info,
-    const AliasInfo& alias_info,
+KernelArgumentHolder allocateOutputs(
+    const Fusion* fusion,
+    const std::vector<GlobalBufferInfo>& output_infos,
+    const std::vector<int>& output_alias_to_input_map,
     const c10::Device& device,
-    ExpressionEvaluator& ee) {
-  FUSER_PERF_SCOPE("fusion_executor::allocations::allocateTensor");
-  // Handle a fusion with duplicated outputs.
-  TensorView* out_tv = out_info.tv;
-  if (ee.isKnown(out_tv)) {
-    return ee.evaluate(out_tv).as<at::Tensor>();
-  }
-
-  std::optional<at::Tensor> aliased_io_tensor = std::nullopt;
-  Val* aliased_io = alias_info.aliased_io;
-  if (aliased_io != nullptr) {
-    NVF_ERROR(
-        aliased_io->isFusionInput() || aliased_io->isFusionOutput(),
-        aliased_io->toInlineString(),
-        " is expected to be a fusion input/output. `ee.evaluate` ",
-        "an intermediate tensor may involve GPU computation to materialize it ",
-        "to global memory.");
-    const PolymorphicValue& aliased_io_val = ee.evaluate(aliased_io);
-    NVF_ERROR(
-        aliased_io_val.is<at::Tensor>(),
-        "Alias io only supports tensor. Found ",
-        PolymorphicValue_functions::toString(aliased_io_val));
-    aliased_io_tensor = aliased_io_val.as<at::Tensor>();
-  }
+    const KernelArgumentHolder& args,
+    bool dynamic_evaluate) {
+  FUSER_PERF_SCOPE("fusion_executor::allocations::allocateOutputs");
 
-  switch (alias_info.type) {
-    case AllocationType::New: {
+  KernelArgumentHolder out_tensors;
+  out_tensors.resize(output_infos.size());
+  for (auto out_idx : c10::irange(output_infos.size())) {
+    auto out_info = output_infos.at(out_idx);
+    if (output_alias_to_input_map.at(out_idx) == -1) {
```

Comment:
(I must be missing something because the PR indeed passes related tests)

The output of a MatmulOp will have AllocateType::Evaluate and aliased_io == nullptr. Doesn't this lead to `output_alias_to_input_map.at(output_idx) == -1` and therefore an empty, incorrect output tensor?

Reply:
I'll check, but maybe it's because anything that's purely evaluated should be run through ExprEvalExecutor, not KernelExecutor. I don't think we can generate a kernel with outputs aliased to one another, but could be missing something. I'll take a look at the test you mention.

Reply:
Yes, this is correct. The reshape/permute in the output to output alias test is being run through ExprEvalExecutor not through KernelExecutor. KernelExecutor only sees the add op from that example. ExprEvalExecutor is completely separate logic and is fine with output-output aliases.